JS: Improve stability of JS model building (prevent cycles)

matthiasblaesing opened the pull request:
As analyzed by @akronenw certain JS files can lead the JS model generator of NetBeans
to create a cyclic structure. The structure looks like this:

```javascript
var Base = new function() {
     // a lot of code

    function Base() {
        // some code
    }

    // more code
}
```

This is turn is at least partially responsible for seemingly endless scanning.

The implemented fix added an assert that detects cycles in the generated model and fixes the
one detected when the `paper-full.js` file is parsed.

In addition a wrong assertion in the GraalJS fork is disabled, that triggered on angular projects
and prevented usage of NetBeans with activated assertions there.

This is not a full fix, but should improve the situation and make it easier to find other
problems in that area by just using NetBeans with assertions enabled (for example by running
a nightly build).

Closes: #4745

matthiasblaesing commented:
This is build on top of #5765, which has to be merged first.

junichi11 commented on webcommon/javascript2.model/src/org/netbeans/modules/javascript2/model/JsObjectImpl.java:
```diff
@@ -192,7 +191,17 @@ public JsObject getParent() {
+        boolean assertionsEnabled = false;
+        assert assertionsEnabled = true;
+        if (assertionsEnabled) {
+            for(JsObject checkParent = newParent; checkParent != null; checkParent = checkParent.getParent()) {
+                if(checkParent == this) {
+                    throw new AssertionError("Cycle detected in " + getFileObject().getPath());
+                }
+            }
```
Nitpick: formatting

junichi11 reviewed (approved):
Great work!
I've verified that `Base()` is shown in the Navigator.
Before:
![nb-gh-5786-before](https://user-images.githubusercontent.com/738383/230800175-38b78b8f-bc79-4dad-aef5-7ae15f55338d.png)
After:
![nb-gh-5786-after](https://user-images.githubusercontent.com/738383/230800183-cf3ae5f0-e235-4aba-8bf7-dca587b5b743.png)
Maybe, we can also add a unit test to `JsStructureScannerTest` of `javascript2.editor` although I'm not sure about the JS area.

matthiasblaesing commented:
@junichi11 thank you for the review. You are right in both points and I updated this PR in place. The formatting as updated using NetBeans formatter and I added a test, that tests the detected structure. This adds a second perspective to the one already checked via `ModelTest#testObjectNameMatchingNestedFunction`, which might be helpful later, when updates are necessary.

junichi11 commented:
Perfect!

matthiasblaesing commented:
@junichi11 thank you for review,  @akronenw thank you for testing. I'll merge with all test runs being green.
